src/pipeline.cc

```diff
@@ -430,4 +430,122 @@ Result Pipeline::UpdateOpenCLBufferBindings() {
+
+  // For each SET command, do the following:
+  // 1. Find the descriptor map entry for that argument.
+  // 2. Find or create the buffer for the descriptor set and binding pair.
+  // 3. Write the data for the SET command at the right offset.
+  for (const auto& arg_info : SetArgValues()) {
+    uint32_t descriptor_set = std::numeric_limits<uint32_t>::max();
+    uint32_t binding = std::numeric_limits<uint32_t>::max();
+    uint32_t offset = 0;
+    uint32_t arg_size = 0;
+    bool uses_name = !arg_info.name.empty();
+    Pipeline::ShaderInfo::DescriptorMapEntry::Kind kind =
+        Pipeline::ShaderInfo::DescriptorMapEntry::Kind::POD;
+    for (const auto& entry : iter->second) {
+      if (entry.kind != Pipeline::ShaderInfo::DescriptorMapEntry::Kind::POD &&
+          entry.kind !=
+              Pipeline::ShaderInfo::DescriptorMapEntry::Kind::POD_UBO) {
+        continue;
+      }
+
+      // Found the right entry.
+      if (entry.arg_name == arg_info.name ||
+          entry.arg_ordinal == arg_info.ordinal) {
+        descriptor_set = entry.descriptor_set;
+        binding = entry.binding;
+        offset = entry.pod_offset;
+        arg_size = entry.pod_arg_size;
+        kind = entry.kind;
+        break;
+      }
+    }
+
+    if (descriptor_set == std::numeric_limits<uint32_t>::max() ||
+        binding == std::numeric_limits<uint32_t>::max()) {
+      std::string message =
+          "could not find descriptor map entry for SET command: kernel " +
+          shader_info.GetEntryPoint();
+      if (uses_name) {
+        message += ", name " + arg_info.name;
+      } else {
+        message += ", number " + std::to_string(arg_info.ordinal);
+      }
+      return Result(message);
+    }
+
+    auto buf_iter = opencl_pod_buffer_map_.lower_bound(
+        std::make_pair(descriptor_set, binding));
+    Buffer* buffer = nullptr;
+    if (buf_iter == opencl_pod_buffer_map_.end() ||
+        buf_iter->first.first != descriptor_set ||
+        buf_iter->first.second != binding) {
+      // Ensure no buffer was previously bound for this descriptor set and
+      // binding pair.
+      for (const auto& buf_info : GetBuffers()) {
+        if (buf_info.descriptor_set == descriptor_set &&
+            buf_info.binding == binding)
+          return Result("previously bound buffer " +
+                        buf_info.buffer->GetName() +
+                        " to PoD args at descriptor set " +
+                        std::to_string(descriptor_set) + " binding " +
+                        std::to_string(binding));
+      }
+
+      // Add a new buffer for this descriptor set and binding.
+      opencl_pod_buffers_.push_back(MakeUnique<Buffer>());
+      buffer = opencl_pod_buffers_.back().get();
+      buffer->SetBufferType(
+          kind == Pipeline::ShaderInfo::DescriptorMapEntry::Kind::POD
+              ? BufferType::kStorage
+              : BufferType::kUniform);
+      DatumType char_type;
+      char_type.SetType(DataType::kUint8);
```

Comment:
Is the point here that it doesn't matter what we set here, because it gets overwritten later with whatever format the data is?

Reply:
It is also the most convenient way to calculate the resizing versus the data in the descriptor map. Everything in the descriptor map is byte-based so this also simpler code (e.g the called to `ResizeTo`).